crates/dojo-erc/src/erc1155/erc1155.cairo

```diff
@@ -0,0 +1,449 @@
+    use dojo_erc::erc1155::components::{ Balance, OperatorApproval, Uri };
+
+    const UNLIMITED_ALLOWANCE: felt252 = 3618502788666131213697322783095070105623107215331596699973092056135872020480;
+
+
+    // Account
+    const IACCOUNT_ID: u32 = 0xa66bd575_u32;
+    // ERC 165 interface codes
+    const INTERFACE_ERC165: u32 = 0x01ffc9a7_u32;
+    const INTERFACE_ERC1155: u32 = 0xd9b67a26_u32;
+    const INTERFACE_ERC1155_METADATA: u32 = 0x0e89341c_u32;
+    const INTERFACE_ERC1155_RECEIVER: u32 = 0x4e2312e0_u32;
+    const ON_ERC1155_RECEIVED_SELECTOR: u32 = 0xf23a6e61_u32;
+    const ON_ERC1155_BATCH_RECEIVED_SELECTOR: u32 = 0xbc197c81_u32;
+
+    #[starknet::interface]
+    trait IERC1155TokenReceiver {
+        fn on_erc1155_received(
+            self: ContractState,
+            operator: ContractAddress,
+            from: ContractAddress,
+            token_id: u256,
+            amount: u256,
+            data: Array<u8>
+        ) -> u32;
+        fn on_erc1155_batch_received(
+            self: ContractState,
+            operator: ContractAddress,
+            from: ContractAddress,
+            token_ids: Array<u256>,
+            amounts: Array<u256>,
+            data: Array<u8>
+        ) -> u32;
+    }
+
+    #[starknet::interface]
+    trait IERC165 {
+        fn supports_interface(self: ContractState, interface_id: u32) -> bool;
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        TransferSingle: TransferSingle,
+        TransferBatch: TransferBatch,
+        ApprovalForAll: ApprovalForAll
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TransferSingle {
+        operator: ContractAddress,
+        from: ContractAddress,
+        to: ContractAddress,
+        id: u256,
+        value: u256
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct TransferBatch {
+        operator: ContractAddress,
+        from: ContractAddress,
+        to: ContractAddress,
+        ids: Array<u256>,
+        values: Array<u256>
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct ApprovalForAll {
+        account: ContractAddress,
+        operator: ContractAddress,
+        approved: bool
+    }
+
+    #[storage]
+    struct Storage {
+        world_address: ContractAddress,
+    }
+
+    //
+    // Constructor
+    //
+
+    #[constructor]
+    fn constructor(ref self: ContractState, world: ContractAddress, uri: felt252) {
+        self.world_address.write(world);
+        _set_uri(ref self, uri);
+    }
+
+    #[view]
+    fn supports_interface(interface_id: u32) -> bool {
+        interface_id == INTERFACE_ERC165 ||
+        interface_id == INTERFACE_ERC1155 ||
+        interface_id == INTERFACE_ERC1155_METADATA
+    }
+
+    //
+    // ERC1155Metadata
+    //
+
+    #[view]
+    fn uri(ref self: ContractState, token_id: u256) -> felt252 {
+        let token = get_contract_address();
+        let query: Query = (token, (u256_as_allowance(token_id))).into_partitioned();
```

Comment:
The confusing `u256_as_allowance` can now be replaced by `token_id.try_into().unwrap()`. Same below in `balance_of`.

Reply:
Do you know a way of keeping the try_into() inlined ?

I'm getting this error:

```
error: Type annotations needed. Failed to infer ?1
 --> erc1155.cairo:132:46
        let query: Query = (token, (account, id.try_into().unwrap())).into_partitioned();
                                             ^***********^
```
I'm solving it like that:
```
let id_felt: felt252 = id.try_into().unwrap();
let query: Query = (token, (account, id_felt)).into_partitioned();
 ```
 Is there a cleaner way ?

Reply:
I don't think there is :(

We could add a `try_into_partitioned` trait to `query.cairo` to deal with `u256` based IDs.